sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVFileFormat.scala

```diff
@@ -174,13 +174,18 @@ private[csv] class CsvOutputWriter(
     context: TaskAttemptContext,
     params: CSVOptions) extends OutputWriter with Logging {
 
-  private val charset = Charset.forName(params.charset)
+  private var univocityGenerator: Option[UnivocityGenerator] = None
 
-  private val writer = CodecStreams.createOutputStreamWriter(context, new Path(path), charset)
-
-  private val gen = new UnivocityGenerator(dataSchema, writer, params)
+  override def write(row: InternalRow): Unit = {
+    val gen = univocityGenerator.getOrElse {
```

Comment:
Also, one thing we should not forget about is, CSV _could_ have headers even if the records are empty.

Reply:
I do think it is fine to write only headers if an user wants to have them. Filtering the header out on this level could be slightly difficult.